web/index.html

```diff
@@ -552,64 +601,147 @@ <h3>Used in</h3>
+                        value: element
+                    };
+                    let craftingList = [tree];
+                    while (craftingList.length > 0) {
+                        let currentNode = craftingList.shift();
+                        let currentElement = currentNode.value;
+
+                        if (usedElements.has(currentElement)) {
+                            continue;
+                        }
+
+                        if (this.saveData.size == 0 || this.showFullRecipe) {
+                            currentNode.left = { value: this.data.craftingCache[currentElement][0] };
+                            currentNode.right = { value: this.data.craftingCache[currentElement][1] };
+                            usedElements.add(currentElement);
+                        } else {
+                            if (!this.data.data[currentElement] && currentElement != "") {
+                                await this.loadChunk(this.getChunk(currentElement));
+                            }
+                            let fromIndex = null;
+                            for (let i = currentNode.lastIndex != undefined ? currentNode.lastIndex + 1 : 0; i < this.data.data[currentElement].from.length; i++) {
+                                if (this.saveData.has(this.data.data[currentElement].from[i][0])) {
+                                    if (this.saveData.has(this.data.data[currentElement].from[i][1])) {
+                                        fromIndex = i;
+                                        break;
+                                    }
+                                    if (!usedElements.has(this.data.data[currentElement].from[i][1])) {
+                                        fromIndex = i;
+                                        break;
+                                    }
+                                } else if (this.saveData.has(this.data.data[currentElement].from[i][1])) {
+                                    if (!usedElements.has(this.data.data[currentElement].from[i][0])) {
+                                        fromIndex = i;
+                                        break;
+                                    }
                                 } else {
-                                    hiddenTo = hiddenTo.split(sep2).map(x => x.split(sep3))
+                                    fromIndex = i;
+                                    break;
                                 }
-                                return [x[0], { from, to, hiddenFrom, hiddenTo }]
-                            }))
-                            this.data.data = Object.freeze(Object.assign({}, this.data.data, chunkData))
-                            this.data.metadata.loadedChunks.push(chunk)
-                        });
+                            }
+
+                            if (fromIndex == null) {
+                                console.log('Added  looping element: ' + this.data.index[currentElement][1]);
```

Comment:
If you want me to explain how all of this works including the dependency loop checks, then I'm happy to do that, but that can happen on discord not in a million comments lol

Reply:
the main gist is that instead of constructing the crafting recipe as a list, we construct it fully as a tree, this means that if we ever reach a recipe that we are supposed to craft that has no options left, we can simply pop back to the last branch of that tree and try again. This also does 99% of the stuff we need when implementing the tree view later on